drivers/source/UnbufferedSerial.cpp

```diff
@@ -0,0 +1,108 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "drivers/UnbufferedSerial.h"
+
+#if DEVICE_SERIAL
+
+#include "platform/mbed_critical.h"
+
+namespace mbed {
+
+UnbufferedSerial::UnbufferedSerial(
+    PinName tx,
+    PinName rx,
+    int baud
+) : SerialBase(tx, rx, baud)
+{
+    // No lock needed in the constructor
+}
+
+ssize_t UnbufferedSerial::write(const void *buffer, size_t size)
+{
+    const unsigned char *buf = static_cast<const unsigned char *>(buffer);
+
+    if (size == 0) {
+        return 0;
+    }
+
+    bool lock_api = !core_util_in_critical_section();
+
+    if (lock_api) {
+        api_lock();
+    }
+
+    for (size_t i = 0; i < size; i++) {
+        _base_putc(buf[i]);
+    }
+
+    if (lock_api) {
+        api_unlock();
+    }
+
+    return size;
+}
+
+ssize_t UnbufferedSerial::read(void *buffer, size_t size)
+{
+    unsigned char *buf = static_cast<unsigned char *>(buffer);
+    if (size == 0) {
+        return 0;
+    }
+    buf[0] = _base_getc();
+    return 1;
+}
+
+short UnbufferedSerial::poll(short events)
+{
+    short revents = 0;
+    if ((events & POLLIN) && SerialBase::readable()) {
+        revents |= POLLIN;
+    }
+    if ((events & POLLOUT) && SerialBase::writeable()) {
+        revents |= POLLOUT;
+    }
+    return revents;
+}
+
+/** Acquire exclusive access to this serial port
+ */
+void UnbufferedSerial::lock()
+{
+    // No lock used - external synchronization required
+}
+
+/** Release exclusive access to this serial port
+ */
+void UnbufferedSerial::unlock()
+{
+    // No lock used - external synchronization required
+}
+
+void UnbufferedSerial::api_lock(void)
+{
+    _mutex.lock();
```

Comment:
For this "raw" class the `lock`/`api_lock` distinction isn't fruitful. You can just have the base `lock`. UARTSerial separated deactivated `lock` cos it knew it was in full control of synchronisation, and would be making `SerialBase` calls from IRQ.

The question is whether `UnbufferedSerial::lock()` should a) do nothing; b) be a `PlatformMutex` claim; or c) be a conditional `PlatformMutex` claim. 

Most driver classes do (b), and a couple do (a).

Basically people have been using HAL serial ports without locks for ages now, via either `RawSerial` or the raw console, so I don't think sticking mutexes around the HAL call is buying is much if anything. I think I'd just put an empty `lock()` in. @bulislaw?

Reply:
It make sense to be lockless as long as it's clear from the docs.

Reply:
[Done](https://github.com/ARMmbed/mbed-os/compare/707656ffd12e37cbe324228bcf83e327a1ae0e10..ed45befc9a6f933a9c2a5b4d6781874d9013df79).